Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert)

mwickert opened the pull request:
First draft submitted  for review.

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/5

Carreau commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,465 @@
+   augment ``scipy.signal`` in a very positive way and enable rapid prototyping of
+   communications and signal processing algorithms. Both student and industry team
+   members in subcontract work, have responded favorably to the use of Python as an
+   engineering problem solving platform. In teaching IPython notebooks are used to augment
```
I think you might want a comma after teaching here. Otherwise I read it as if you are teaching IPython Notebook itself.

mwickert replied:
Matthias,

Thanks for this catch. I applied the change.

Mark

From: Matthias Bussonnier [mailto:notifications@github.com]
Sent: Sunday, June 21, 2015 8:29 AM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


In papers/mark_wickert/mark_wickert.rst<https://github.com/scipy-conference/scipy_proceedings/pull/157#discussion_r32893928>:

> +.. class:: abstract

> +

> +   This paper will take the audience through the story of how an electrical and computer

> +   engineering faculty member has come to embrace Python, in particular IPython Notebook

> +   (IPython kernel for Jupyter),

> +   as an analysis and simulation tool for both teaching and research in signal processing

> +   and communications. Legacy tools such as MATLAB are well established (entrenched) in

> +   this discipline, but engineers need to be aware of alternatives, especially in the case

> +   of Python where there is such a vibrant community of developers and a seemingly well

> +   designed language. In this paper case studies will also be used to describe domain

> +   specific code modules that are being developed to support both lecture and lab oriented

> +   courses going through the conversion from MATLAB to Python. These modules in particular

> +   augment ``scipy.signal`` in a very positive way and enable rapid prototyping of

> +   communications and signal processing algorithms. Both student and industry team

> +   members in subcontract work, have responded favorably to the use of Python as an

> +   engineering problem solving platform. In teaching IPython notebooks are used to augment

I think you might want a comma after teaching here. Otherwise I read it as if you are teaching IPython Notebook itself.

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157/files#r32893928>.

Carreau commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,465 @@
+
+The next phase of the journey, focuses on the research and development
+side of signals and systems work. During a sabbatical working for a small engineering firm, academic year
+2013--2014, Python and IPython notebook (IPython kernel for Jupyter) served as the primary digital signal
```
Nitpicking, you have 2 dashes here, and one on line 63. If you prefer to be consistent.

mwickert replied:
Mathias,

Thanks for this catch too. I was thinking LaTeX and wanting an emdash. A dash is OK too.

Mark

From: Matthias Bussonnier [mailto:notifications@github.com]
Sent: Sunday, June 21, 2015 8:38 AM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


In papers/mark_wickert/mark_wickert.rst<https://github.com/scipy-conference/scipy_proceedings/pull/157#discussion_r32893999>:

> +Introduction

> +------------

> +

> +This journey into Python for electrical engineering problem solving began with

> +the writing of the book *Signals and Systems for Dummies* [Wic2013]_, published summer

> +2013. This book features the use of Python (``Pylab``) to

> +bring life to the mathematics behind signals and systems theory. Using Python in the Dummies book was done to make

> +it easy for all readers of the book to develop their signals and system problem solving skills, without additional

> +software tools investment. Additionally, the author provides the custom code module ``ssd.py`` [ssd]_, which is built on

> +top of  ``numpy``,  ``matplotlib``, and ``scipy.signal``, to make it easy to work and extend the examples found in

> +the book. Engineers love to visualize their work with plots of various types. All of the plots in the book were

> +created using Python, specifically ``matplotlib``.

> +

> +The next phase of the journey, focuses on the research and development

> +side of signals and systems work. During a sabbatical working for a small engineering firm, academic year

> +2013--2014, Python and IPython notebook (IPython kernel for Jupyter) served as the primary digital signal

Nitpicking, you have 2 dashes here, and one on line 63. If you prefer to be consistent.

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157/files#r32893999>.

Carreau commented:
the link to ssd.py (and docs) seem to timeout for me. May I suggest putting a copy on github ?
You seem to be new to github, so feel free to ask if you need a hand. 

The modules do not seem to be on PyPI either, even during developement, I believe it would be e good thing to do, but I guess you might have restriction from the book publisher ?

Carreau commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,465 @@
+The present lecturing style for all courses I teach
+involves the use of a tablet PC, a data projector, a microphone, and audio/video screen capture software, e.g.
+Camtasia Studio [Camtasia]_. Live Python demos are run in the notebook, and in many cases all the code is developed in
+real-time as questions come from the class. The notebook is more that just a visual experience.
```
I think you mean "is more **than** just"

mwickert replied:
Matthias,

Thanks for this now obvious catch. I found another nearby. These typos can be hard to find.

Mark

From: Matthias Bussonnier [mailto:notifications@github.com]
Sent: Sunday, June 21, 2015 8:51 AM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


In papers/mark_wickert/mark_wickert.rst<https://github.com/scipy-conference/scipy_proceedings/pull/157#discussion_r32894097>:

> +tidy, but when introducing new concepts to students, placing algorithms inside notebooks has the advantage of making

> +the code visible, and invites tinkering.

> +

> +Describing IPython Notebook Use Cases

> +-------------------------------------

> +

> +In this section I describe how Python and in particular the IPython notebook has been integrated into teaching,

> +graduate student research, and industry research and development.

> +

> +Teaching

> +========

> +

> +The present lecturing style for all courses I teach

> +involves the use of a tablet PC, a data projector, a microphone, and audio/video screen capture software, e.g.

> +Camtasia Studio [Camtasia]_. Live Python demos are run in the notebook, and in many cases all the code is developed in

> +real-time as questions come from the class. The notebook is more that just a visual experience.

I think you mean "is more than just"

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157/files#r32894097>.

Carreau commented:
Some domain specific term (survivor traceback paths , the trellis back, the decision depth) are not well explained, but this is not the goal of the article, so it does not prevent comprehension.

Next to that I have 2 question remarks, which are not in any case requests to change the way things are.

First, in the screenshot, It seems like you are not following pep8, I know that "A Foolish Consistency is the Hobgoblin of Little Minds" as the original pep mention, but would it be good to give good habits to students ?

Second, from experience I found that using pylab to draw figures was problematic as it is ment to mimick matlab, and beginners tends to have difficulties migrating to matplotlib object-oriented way later. While if you teach the OO way since the beginning, it does not seem to be more difficult. Do you have thoughts on that ?

mwickert commented:
Matthias

I was traveling over the weekend. I have started to look at your comments. I have not looked at the pep8 document for a while, but have scammed through it again tonight. I will response to your other comments soon.

Prior to hearing from you I added another case study and enhanced another one. I will try to do a push in the morning here in Colorado. I am guessing I am 6-8 hrs time different from you.

Mark Wickert


On Jun 21, 2015, at 9:16 AM, Matthias Bussonnier <notifications@github.com<mailto:notifications@github.com>> wrote:


Some domain specific term (survivor traceback paths , the trellis back, the decision depth) are not well explained, but this is not the goal of the article, so it does not prevent comprehension.

Next to that I have 2 question remarks, which are not in any case requests to change the way things are.

First, in the screenshot, It seems like you are not following pep8, I know that "A Foolish Consistency is the Hobgoblin of Little Minds" as the original pep mention, but would it be good to give good habits to students ?

Second, from experience I found that using pylab to draw figures was problematic as it is ment to mimick matlab, and beginners tends to have difficulties migrating to matplotlib object-oriented way later. While if you teach the OO way since the beginning, it does not seem to be more difficult. Do you have thoughts on that ?

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157#issuecomment-113912686>.

Carreau commented:
> I was traveling over the weekend. I have started to look at your comments. I have not looked at the pep8 document for a while, but have scammed through it again tonight. I will response to your other comments soon.

No worries, after a bit the missing space after commas jump to your eyes, but it's ok. 

> Prior to hearing from you I added another case study and enhanced another one. I will try to do a push in the morning here in Colorado. I am guessing I am 6-8 hrs time different from you.

I'll try to re-read soon, I am traveling so have weirds schedule. 

-- 
Matthias

mwickert commented:
Matthias,

Your suggestions on github or PyPI are good. I have no restrictions from my publisher. I simply have not taken the initiative, and for this paper signed obtain a github account. In my consulting work I am now using git regularly, so I should do it write. Maybe during the SciPy conference I will get this going.

In the mean time a ZIP of the modules I have talked about in the paper is here: http://www.eas.uccs.edu/wickert/ece5625/lecture_notes/Python_Modules.zip . The code is not pep8, but ssd.py does have numpy docstrings and Sphinx documentation is here: http://www.eas.uccs.edu/wickert/SSD/docs/python/ .

Mark

From: Matthias Bussonnier [mailto:notifications@github.com]
Sent: Sunday, June 21, 2015 8:47 AM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


the link to ssd.py (and docs) seem to timeout for me. May I suggest putting a copy on github ?
You seem to be new to github, so feel free to ask if you need a hand.

The modules do not seem to be on PyPI either, even during developement, I believe it would be e good thing to do, but I guess you might have restriction from the book publisher ?

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157#issuecomment-113910817>.

Carreau commented:
> Your suggestions on github or PyPI are good. I have no restrictions from my publisher. I simply have not taken the initiative, and for this paper signed obtain a github account. In my consulting work I am now using git regularly, so I should do it write. Maybe during the SciPy conference I will get this going.

Sure, feel free to ask for help. You'll see that more often that you imagine people will just come along and send you PRs. I also find Travis-CI really useful, you just push your code on github, and it does the unit-testing for you, which simplify workflow a lot. PyPI is not that scary, I personally can never remember the step and found out [flit](https://github.com/takluyver/flit) which make my life easier, but which is still a toy project. 

> In the mean time a ZIP of the modules I have talked about in the paper is here: http://www.eas.uccs.edu/wickert/ece5625/lecture_notes/Python_Modules.zip . The code is not pep8, but ssd.py does have numpy docstrings and Sphinx documentation is here: http://www.eas.uccs.edu/wickert/SSD/docs/python/ .

Great, I've had a look and it looks perfect. 

One last tip, when you have your project on GitHub, you can use https://zenodo.org/ to freeze one version and get a DOI which make it citable, but again we can see that durign the conference, 

Thanks.

Carreau commented:
@katyhuff @jaberg I think the paper is ready to be merged.

mwickert commented:
Matthias,

I have applied the pep8 fixes, explained in more detail the digital comm code case study.

I have also added a paragraph explaining why I prefer to use the pylab interface to matplotlib.

I suspect more typos exist, but for now I am done. I have just done a push.

Mark

From: Matthias Bussonnier [mailto:notifications@github.com]
Sent: Sunday, June 21, 2015 9:16 AM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


Some domain specific term (survivor traceback paths , the trellis back, the decision depth) are not well explained, but this is not the goal of the article, so it does not prevent comprehension.

Next to that I have 2 question remarks, which are not in any case requests to change the way things are.

First, in the screenshot, It seems like you are not following pep8, I know that "A Foolish Consistency is the Hobgoblin of Little Minds" as the original pep mention, but would it be good to give good habits to students ?

Second, from experience I found that using pylab to draw figures was problematic as it is ment to mimick matlab, and beginners tends to have difficulties migrating to matplotlib object-oriented way later. While if you teach the OO way since the beginning, it does not seem to be more difficult. Do you have thoughts on that ?

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157#issuecomment-113912686>.

dalippa commented:
Hi @mwickert, I was asked to join the review of your paper. I've finished reading and have some suggestions/comments:

* Abstract is a bit long. Some sentences in the abstract are fragments. Abstract probably could do without the "side note".

* The last 2-3 paragraphs of the introduction are a bit convoluted. Intro also has some dangling references in the introduction make it less clear.

* There is opinion stated in the paper, without referencing facts/citations/evidence (co-worker discussion; "In my view Python with pylab and sympy...")

* Rather than listing the modules of "ssd" and others, it might be helpful to provide a few key examples and where they would be used in signal processing, such as the fir_iir_notch or soi_snoi_gen. The function names seem a bit obscure: a Google search for "NRZ_bits2" showed up with only 3 results, all in the ssd documentation.

* The section "Arriving at Python for Communications and Signal Processing Modeling" is making a few assumptions that the reader has signal processing experience. This seems a particularly steep change from the earlier sections of the paper that were far less technical.

* The paragraph starting "For each new chapter of material from the course text, a new
notebook is created..." is a bit confusing to me.

* The paper style varies in its level of formality from section to section. Was that your intent?

* Paper switches narrative styles from first person, to third person, to second person in the "Describing IPython Notebook Use Cases" section. This makes this section somewhat difficult to follow.

* Under "Fig 4" I believe you have an errant sentence fragment "What to say here".

* The plots and IPython notebooks are very helpful and are very well explained.

* Paper seems a bit on the long side.

mwickert commented:
Dalippa and other reviewers, including @katyhuff and @jaberg

I have worked through suggestions/comments I received June 27. I am ready to call this done pending and more comments. Regarding the 11 bullets, I am providing a written summary of how I addressed each comment.

Bullet 1: I trimmed out several sentences and in particular removed the “side note”.

Bullet2: I worked through readability issues with the introduction. I had a graduate student mark up the entire document. I found additional issues on my own too.

Bullet 3: I removed the “In my view…” sentence. Including this was a bad call on my part.

Bullet 4: I have not removed the module content listing as I feel that they show the enhancements I have added to support teaching in my domain with Python. If shrinking the paper size is important I can omit these lists. The suggested alternative of providing a few key examples to me seems redundant given that the paper contains a case studies section. I would expect functions from ssd.py to appear outside the documentation as the students being exposed to be publishing. I my consulting work other engineers are using these functions, but again are not publishing anything beyond internal documents.

Bullet 5: To me the section on “Arriving at Python ..” is not too technical, but rather still at a high level. I admit in the “Case Studies” section the discussions turns to domain based topics.

Bullet 6: The paragraph “For each new chapter…” is fixed. Thanks for pointing this out.

Bullet 7: My intent with the paper style is to be mostly informal, with the exception of some of the discussion in the Case Studies section. I look at this paper as my attempt at allowing the general scientific/engineering computing community a view into communications and signal processing numerical computations. Of course this view is through the IPython notebook.

Bullet 8: I have tried fix the “Use Cases” section to hace a consistent first person style. Please let me know if I failed to fix something you noticed.

Bullet 9: Sad to say this was a place holder when I was outlining the paper that never got removed.

Bullet 10: Thank you!

Bullet 11: Yes, it now fills 8 pages. I did not think it would get so long when I first started writing. I could take out some of the case studies. I removed a few sentences here and there, but have added a bit more explanation in the case studies, so the overall length has remained about the same.

I did a push just before I wrote this email.

Mark Wickert

From: dalippa [mailto:notifications@github.com]
Sent: Saturday, June 27, 2015 6:33 PM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


Hi @mwickert<https://github.com/mwickert>, I was asked to join the review of your paper. I've finished reading and have some suggestions/comments:

·        Abstract is a bit long. Some sentences in the abstract are fragments. Abstract probably could do without the "side note".

·        The last 2-3 paragraphs of the introduction are a bit convoluted. Intro also has some dangling references in the introduction make it less clear.

·        There is opinion stated in the paper, without referencing facts/citations/evidence (co-worker discussion; "In my view Python with pylab and sympy...")

·        Rather than listing the modules of "ssd" and others, it might be helpful to provide a few key examples and where they would be used in signal processing, such as the fir_iir_notch or soi_snoi_gen. The function names seem a bit obscure: a Google search for "NRZ_bits2" showed up with only 3 results, all in the ssd documentation.

·        The section "Arriving at Python for Communications and Signal Processing Modeling" is making a few assumptions that the reader has signal processing experience. This seems a particularly steep change from the earlier sections of the paper that were far less technical.

·        The paragraph starting "For each new chapter of material from the course text, a new
notebook is created..." is a bit confusing to me.

·        The paper style varies in its level of formality from section to section. Was that your intent?

·        Paper switches narrative styles from first person, to third person, to second person in the "Describing IPython Notebook Use Cases" section. This makes this section somewhat difficult to follow.

·        Under "Fig 4" I believe you have an errant sentence fragment "What to say here".

·        The plots and IPython notebooks are very helpful and are very well explained.

·        Paper seems a bit on the long side.

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157#issuecomment-116172411>.

mwickert commented:
@katyhuff and @jaberg I am ready for the next step. The PDF paper builds fine on my end.

dalippa commented:
@mwickert I have re-reviewed the paper and have the following feedback:

Bullet 1: Resolved

Bullet 2: Introduction (and other sections) still switche tenses within the same paragraphs; tense should be consistent within and across sections. Sentences in 2nd paragraph of intro are still quite long with some run-on sentences. Consider using footnotes to help improve the flow in places (for example: "During a sabbaical working for a small enginering firm, academic year 2013-2014..."; citation of Mathematica and Maxima.

Bullet 3: The discussion with coworkers is still there, without mentioning who they are, their expertise, etc. which is relevant to the claim.

Bullet 4: To the review criteria of "Due to the interdisciplinary nature of SciPy, highly domain-specific jargon should be avoided or explained where possible", there needs to be some clarification on what these functions are. The name of the functions breaks up the flow of the "Modules Developed or Under Development" section and might be better suited in an appendix. The other lists in the section better since they are smaller and break the flow less. Rather than code snippets, you might want to consider using figures/captions  to improve the flow with figure citations.

Bullet 5: I guess my issue with "Arriving at Python for Communications and Signal Processing Modeling" is that the functions are using domain-specific jargon (see Bullet 4). Another example is the function "sinusoidAWGN". Google was helpful in looking that one up, but some of the others seem much more obscure.

Bullet 6: Consider active voice more often since you are the subject of many of the sentences (e.g. "...a new IPython notebook is created..."). Given that the narrative is about teaching style, please consider adding a figure (or citation of a figure in a later section) in case an instructor would like to implement your suggestions of using IPython as a teaching tool. The sentence "When appropriate, starter content is added to say the Chapter x notebook before the lecture, to provide relevant theory right in the notebook and provide continuity between the lecture notes mathematics and the notebook demos" is still confusing and grammatically incorrect. Similar issues in that same paragraph. Also, please consider removing superlatives ("fantastic").

Bullet 7: I think I wasn't clear in my comment. Even though this is an informal paper, sentences like "Since the notebook is live, the inevitable what if questions are asked and hopefully quickly answered" are still awkward. I think the point you're trying to make is that exploration is a necessary part of the teaching and learning process.

Bullet 8: I am still seeing a lot of switches in narrative styles across the entire paper. One such example is in the transition from the first paragraph in "Industry Research and Development" into "Later, prototype algorithm development is started...". It's not clear who is "doing" the action specified. In my opinion, this type of sentence construction impedes the flow of ideas.

Bullet 9: Resolved.

Bullet 10: Resolved.

Bullet 11: I still think 8 pages is a bit long, but if the SciPy folks are good with expanding paper length, I'm good with that.


New comments:
* Since you have a github repository, consider adding footnotes with downloads to pre-built IPython notebooks. I think one important point of the paper is to encourage the use of Python, and pre-built examples should help sell that point. Furthermore, while the diagrams are helpful, your claim in the abstract that tinkering is an important part of the process is *very* true, and examples would lend themselves to that.
* Under the "Case studies", consider changing to active voice in the introductory section, since you are the subject.
* Word usage error in "Case Studies": insure vs ensure.
* Case studies are less fuzzy and more formal; therefore, consider avoiding modal verb that weaken the message of this section (https://en.wikipedia.org/wiki/Modal_verb).
* "Students frequently have problems setting up and evaluating the convolution integral" is a perfect place to cite proof that exploration with the IPython notebook improves learning and understanding. From my perspective, this is the key message of the paper.\
* ". In this same figure the analytical solution is easily plotted for the case of T = 1‘sand : math : ‘a = 5." seems to have some LaTeX formatting issues.
* Consider moving your figures so that they aren't breaking sentence flow of the paper (between paragraphs is ideal).
* Please cite figures consistently (Figures 1-3 are cited differently than the rest)
* The end of the discussion of "ssd.conv_integral()" is a bit awkward and abrupt.
* One thing that isn't clear to me is whether this was used to teach undergraduate or graduate students. Is that be salient?
* "In the real-time digital signal processing course..." section has a few awkward sentences and inconsistent use of font for python packages, which impedes readability. It is less clear to me what the value is in this case of the case study. Consider adding a sentence or two at the beginning of the case to describe the problem that the notebook solves (for students, teaching, and not just the math). Also, the "(5) and (6)" seem to bleed into the next section.
* My biggest issue in the "Statistical Signal Processing" is that it doesn't seem to add to the paper. Consider removing it.
* The conclusion does not seem to mirror the abstract/introduction that the focus is not in IPython as a problem solving tool, but as a means of exploring - for teaching and research. The conclusion seems to weaken and not strengthen your claim. From my perspective there are a few main points that should be reiterated: Python/IPython deserve a place in the classroom and research lab; teaching, learning, and research require exploration and collaboration; though IPython is not widely adopted, it excels at both.

mwickert commented:
Propose making all URLs in the references section RST hyperlinks. Is this desirable?

mwickert commented:
Dallipa,

I have read through your second round of comments. I made a major overhaul to the paper. It would be hard for me to go line-by-line through all of your various comments. I am hoping that via the file diff you will be able to see what I’ve done.

I am also hoping I, with your help, am converging on a ‘solution’.

Major changes include removing the function lists as you originally recommended. I added footnotes (had to review how RST does this again, but the examples helped).

I have left the random signals example in at this point as I feel it is a good example of how a rather complex calculus problem can be easily simulated to allow graduate students to gain confidence in their problem solving skills.

Mark Wickert

From: dalippa [mailto:notifications@github.com]
Sent: Saturday, June 27, 2015 6:33 PM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


Hi @mwickert<https://github.com/mwickert>, I was asked to join the review of your paper. I've finished reading and have some suggestions/comments:

·        Abstract is a bit long. Some sentences in the abstract are fragments. Abstract probably could do without the "side note".

·        The last 2-3 paragraphs of the introduction are a bit convoluted. Intro also has some dangling references in the introduction make it less clear.

·        There is opinion stated in the paper, without referencing facts/citations/evidence (co-worker discussion; "In my view Python with pylab and sympy...")

·        Rather than listing the modules of "ssd" and others, it might be helpful to provide a few key examples and where they would be used in signal processing, such as the fir_iir_notch or soi_snoi_gen. The function names seem a bit obscure: a Google search for "NRZ_bits2" showed up with only 3 results, all in the ssd documentation.

·        The section "Arriving at Python for Communications and Signal Processing Modeling" is making a few assumptions that the reader has signal processing experience. This seems a particularly steep change from the earlier sections of the paper that were far less technical.

·        The paragraph starting "For each new chapter of material from the course text, a new
notebook is created..." is a bit confusing to me.

·        The paper style varies in its level of formality from section to section. Was that your intent?

·        Paper switches narrative styles from first person, to third person, to second person in the "Describing IPython Notebook Use Cases" section. This makes this section somewhat difficult to follow.

·        Under "Fig 4" I believe you have an errant sentence fragment "What to say here".

·        The plots and IPython notebooks are very helpful and are very well explained.

·        Paper seems a bit on the long side.

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157#issuecomment-116172411>.

mwickert commented:
@dallipa, @katyhuff and @jaberg from the above you can see that I completed a major rework of the paper last night. I am running out of time my own schedule. In particular I travel to the conference on Sunday morning, the 5th. Keep me posted on my next next so I can bring my side to completion. Katy, I have not found the status links you mention in your latest Email.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+the writing of the book *Signals and Systems for Dummies* [Wic2013]_, published summer
+2013. This book features the use of Python (``Pylab``) to
+bring life to the mathematics behind signals and systems theory. Using Python in the Dummies book is done to make
+it easy for all readers of the book to develop their signals and system problem solving skills, without additional
```
This sentence is still awkward.

mwickert replied:
May true, but I have decided to leave it as is.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+processing modeling tools on three different projects. Initially it was not clear which tool would be best, but
+following discussions with co-workers [#]_ Python seemed to be the right choice. Note, for the most part, the analysis
+team was new to Python, all of us having spent many years using MATLAB/Octave [MATLAB]_/[Octave]_. A nice motivating
+factor is that Python is already in the workflow in the real-time DSP platform used by the company.
```
I think one of the other major points you make is that it's not just used by the *company* but by the *industry* as a whole - and that Python/IPython can augment the standard toolbox, or disrupt it.

mwickert replied:
I agree, but have not added or deleted any text.

dalippa replied:
OK

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+.. [#] See for example *Mathematica* [Mathematica]_ (commercial) and *wxMaxima* [Maxima]_ (open source).
+
+Getting other faculty on-board is the next step. I am optimistic and
```
This idea doesn't fit here; I think this belongs at the end of the paper - almost like "Future Work," but it could go in the conclusion.

mwickert replied:
Suggestion taken. Now in the Future Work discussion at the end of the paper.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+Getting other faculty on-board is the next step. I am optimistic and
+look forward to this challenge as tutorial sessions are planned over summer 2015.
+
+The remainder of this paper is organized into the following sections: arriving at Python for communications and signal
```
I don't think this sentence is really necessary. Each section should naturally flow, without being explicit.

mwickert replied:
I like having kind of a preview of what is to come in the paper. I am leaving it in.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+open-source alternatives over MATLAB. I initially homed in on using Octave [Octave]_ for its syntax compatibility
+with MATLAB. Later I started to explore Python and became fascinated by the ease of use offered by the IPython (QT)
+console and the high quality of ``matplotlib`` 2D plots. The full power of Python
+for engineering and scientific computing became evermore obvious as I learned more about the language and the
```
This sentence is a bit awkward.

mwickert replied:
I made a few tweaks.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+When I took on the assignment of writing the *Signals and Systems for Dummies* book [Wic2013]_ Python seemed like a
+good choice because of the relative ease with which anyone could obtain the tools and then get hands-on experience with
+the numerical examples I was writing into the book. The power of ``numpy`` and algorithms available in ``scipy`` are
+indeed wonderful, but I immediately recognized that enhancements to ``scipy.signal`` are needed to make signals
```
"The power of numpy ... wonderful" is subjective, and not needed here.

mwickert replied:
Added a few tweaks.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+and systems tinkering user friendly. As examples were written for the book, I began to write support functions that
+fill in some of the missing details not found in ``scipy``. This is the basis for the module ``ssd.py``.
+At present writing additional support modules for signals and systems work (in particular communications and signal
+processing) is ongoing.
```
`"This is the basis for the module ``ssd.py``. At present writing additional support modules for signals and systems work (in particular communications and signal processing) is ongoing."` may be better as: "This is the basis for the module ssd.py, a constant work in progress to make open source signals and systems software more accessible to the engineering community."

mwickert replied:
Accepted suggested change!

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+Modules Developed or Under Development
+======================================
+
+As already briefly mentioned, the first code module I developed is ``ssd.py`` [#]_. This module contains 61 functions
```
Links are ephemeral, github repositories are less so. If it's in github, I recommend using that link if available.

mwickert replied:
Yes, I want to move in this direction. Writing this paper has been  my first exposure to using github. I have been using git in a work setting for a while. The bottom line however is that the publisher of the Dummies book requires that I keep this link in place as it stands now. If I move the site to github then I think I would want to move everything else for the book too, but this requires some thinking through. May you or others can give me some more thoughts at the conference.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+As already briefly mentioned, the first code module I developed is ``ssd.py`` [#]_. This module contains 61 functions
+supporting *signal* generation, manipulation, and display, and *system* generation and characterization. Some of
+the functions implement subsystems such as a ten band audio equalization filter or the moel of an automobile cruise
```
moel? spelling error?

mwickert replied:
Fixed!

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+A third module, ``fec_conv.py``, implements a rate one-half *convolutional encoding* and *decoding* class.
+Arbitrary constraint length codes can be employed as well as *puncturing* and *depuncturing* patterns. For decoding the
+Viterbi algorithm with soft decision metrics is used. The class contains seven methods that include two graphical
```
"For decoding..." is a fragment.

mwickert replied:
This area has been reworked as a background discussion of convolutional coding and decoding is moved from the case study to here. The case study is now streamlined as the background material is in place ahead or time. One of your later comments motivated me to make this change.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+Viterbi algorithm with soft decision metrics is used. The class contains seven methods that include two graphical
+display functions, one of which shows the *survivor traceback paths* through the trellis back to the
+decoder *decision depth*. This graphical display gives students insight into the operation of the Viterbi algorithm,
+which at a high level is a *dynamic programming* algorithm. Besides the class ``fec_conv``, this module also contains
```
"This graphical display..." is still very awkward.

mwickert replied:
This has been rewritten.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+four functions for computing error probability bounds using
+the *weight structure* of the code under both *hard* and *soft* branch metric distance calculations [Zie2015]_.
+
+Both the encoder and especially the Viterbi decoder, are numerically intensive. Speed enhancements, perhaps using
```
This belongs in future work.

mwickert replied:
This has been moved to the Furture work discussion at the end of the paper.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+A fourth module, ``synchronization.py``, was developed while teaching a *phase-locked loops* course, Summer 2014. This
+module supplies eight simulation functions ranging from a basic phase-locked loop and both carrier and symbol synchronization
+functions for digital communications waveforms. Synchronization is extremely important is all modern communications
+communications schemes. This module was also utilized in an analog communications course taught Spring 2015.
```
"Synchronization is extremely important..." needs some more explanation.

mwickert replied:
Fixed by reorganizing the entire paragraph.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+functions for digital communications waveforms. Synchronization is extremely important is all modern communications
+communications schemes. This module was also utilized in an analog communications course taught Spring 2015.
+As a side note, modern communication systems fail to get data bits through a wireless link when
+synchronization fails.
```
Please avoid side notes. Try to integrate the importance of synchronization into the paragraph. This is the motivation behind the development of the module, and should probably be in the first sentence.

mwickert replied:
Eliminated in the above reorganization of the paragraph.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+As a side note, modern communication systems fail to get data bits through a wireless link when
+synchronization fails.
+
+More modules are planned as well as enhancements to the existing modules. A great side benefit of using IPython
```
Please extract future work into a separate section to maximize the flow of this section's ideas.

mwickert replied:
Done.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+tidy, but when introducing new concepts to students, placing algorithms inside notebooks has the advantage of making
+the code visible, and invites tinkering.
+
+Describing IPython Notebook Use Scenarios
```
Perhaps: `IPython Notebook Use Cases In Teaching and Research`?

mwickert replied:
Leaving this as is.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+========
+
+The present lecturing style for all courses I teach
+involves the use of a tablet PC, a data projector, a microphone, and audio/video screen capture software, e.g.
```
Other teaching materials don't seem relevant. What *is* relevant is the additional enhancements that Python/IPython provide in the teaching/learning process.

mwickert replied:
Not removed, but reworked.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+action of systems changes what can be heard. Students enjoy hearing as well as seeing results.
+By interfacing the tablet *lineout* or *headphone* output to the podium
+interface to the classroom speakers, everyone can hear the impact of algorithm tweaks on what is being heard. This
+is where the fun starts! Core modules such as ``ssd.py`` described earlier, are imported at the top of each notebook.
```
Live updates are very salient here. Most people get great value out of visualizing data; providing auditory feedback in addition is very valuable.

"Core modules..." is not necessary.

mwickert replied:
Fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+particular math model relates to real-word communications and signal processing systems.
+
+Computer projects benefit greatly from the use of the notebook, as sample notebooks with starter code are easily
+posted to the course Web Site. The sample notebook serves as a template for the project report document that the
```
This is probably a novel idea: an IPython notebook as a submission tool, in place of a paper notebook. Consider adding a little more on your experience witht hat. Consider removing the PDF conversion/screen shots. *Interactivity* is the primary theme of the paper. PDF/Screen shot conversion, though mentioned in the abstract, is a side note that can probably move elsewhere.

mwickert replied:
I added and deleted some content here.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+In working with graduate students on their research, it is normal to exchange code developed by fellow graduate
+students working on related problems. Explaining how code works with worked examples is a perfect use case for
+IPython notebook. The same approach holds for faculty interaction with their graduate students. In this scenario the
```
"Explaining how..." is awkward.

mwickert replied:
I made some changes here.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+placed in the notebook.
+
+As the student makes progress on a research task they document their work in a notebook. Faculty member(s) are briefed
+on the math models and simulation results. Since the notebook is live, the inevitable *what if* questions are asked and
```
There's another very useful part of the notebook. When used in certain ways, you can see how each individual student is using the notebook (since it shows a history of commands run), and can refine teaching techniques.

mwickert replied:
Accepted suggested change.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+an abstraction of real electronic hardware, in which case a hardware implementer uses the notebook (maybe just a PDF
+version) to create a hardware prototype.
+
+Live From the Classroom
```
This section would flow better with the teaching section. I think a progression from teaching, to research, to industry R&D is a more logical progression.

mwickert replied:
I A am content with the current arrangement.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+=================================
+
+With the notebook engineers working on the same team are able to share analytical models and  development approaches
+using markdown cells. The ability to include equations using LaTeX markup is nice, as mathematical developments,
```
This paragraph is still pretty awkward. Please try to rephrase to avoid less concrete words such as "nice". Maybe: `The inclusion of LaTeX markup is a welcome addition and furthers the establishment of notational conventions...`

mwickert replied:
Accepted suggested change.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+including the establishment of notational conventions, is the first step in the development of signal processing
+algorithms.
+
+Later, prototype algorithm development is started using code cells. Initially synthesized signals (waveforms)
```
Grammatical errors in "Initially synthesized signals..."

mwickert replied:
Fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+are used to validated the core functionality of an algorithm. Next, signal captures from the actual real-time
+hardware are used as a source of test vectors to verify that performance metrics are being achieved. Notebooks
+can again be passed around to team members for further algorithm testing. Soon code cell functions can be moved to
+code modules and the code modules distributed to team members via GIT or some other distributed revision control
```
`GIT` need not be capitalized but can be encapsulated in backticks like other software citations. Please also add citations for matplotlib and git.

mwickert replied:
Done.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+system. At every step of the way ``matplotlib`` graphics are used to visualize performance of a particular
+algorithm, versus say a performance bound.
+
+Complete subsystem testing at the Python level may be sufficient in some cases. In a more typical case, code will
```
I am still finding this paragraph confusing. Is the idea to transition from component-based testing to system-wide testing? If so, that does not seem explicitly stated.

mwickert replied:
Reworked.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+=======================
+
+Here live from the classroom means responding to questions using on-the-fly IPython notebook demos. This is an excellent
+way to show off the power of Python. Sometimes questions come and you feel like building a quick model right then and
```
"This is an excellent way..." could probably be more strongly worded. Perhaps "This is where Python and IPython truly shine: off-the-cuff questions can be answered live in the classroom--solidifying students' understanding."

mwickert replied:
Dallipa,

I appreciate all your feedback, but 26+ today alone is starting overload me. When can I expect this to settle a bit? I have other work to do?

Mark Wickert

From: dalippa [mailto:notifications@github.com]
Sent: Friday, July 3, 2015 7:42 AM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


In papers/mark_wickert/mark_wickert.rst<https://github.com/scipy-conference/scipy_proceedings/pull/157#discussion_r33864390>:

> +hardware are used as a source of test vectors to verify that performance metrics are being achieved. Notebooks

> +can again be passed around to team members for further algorithm testing. Soon code cell functions can be moved to

> +code modules and the code modules distributed to team members via GIT or some other distributed revision control

> +system. At every step of the way ``matplotlib`` graphics are used to visualize performance of a particular

> +algorithm, versus say a performance bound.

> +

> +Complete subsystem testing at the Python level may be sufficient in some cases. In a more typical case, code will

> +be moved to a production environment and recoding may be required. It might also be that the model is simply

> +an abstraction of real electronic hardware, in which case a hardware implementer uses the notebook (maybe just a PDF

> +version) to create a hardware prototype.

> +

> +Live From the Classroom

> +=======================

> +

> +Here live from the classroom means responding to questions using on-the-fly IPython notebook demos. This is an excellent

> +way to show off the power of Python. Sometimes questions come and you feel like building a quick model right then and

"This is an excellent way..." could probably be more strongly worded. Perhaps "This is where Python and IPython truly shine: off-the-cuff questions can be answered live in the classroom--solidifying students' understanding."

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157/files#r33864390>.

mwickert replied:
In the previous comment, I am content to leave it as is.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+Case Studies
+------------
+
+In this section several case studies are presented. Each case study details one or more of the IPython notebook
```
This paragraph can be made more concise.

mwickert replied:
Fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+In all of the case studies you will see that graphical results are produced using the ``pylab`` interface to ``matplotlib``.
+This is done purposefully for two reasons. The first stems from the fact that currently all students have received
+exposure to MATLAB in a prior course, and secondly, to help ensure that returning to MATLAB from Python is easier. Why
+would anyone want to return you might ask? Job requirements may require this and presently MATLAB is pervasive in
```
The question here is less relevant. What is most relevant in this paragraph is that (a) all students have had prior MATLAB training, and (b) you wish to augment, and not replace, the students' MATLAB knowledge since industry is still lagging when it comes to using open source tools.

mwickert replied:
Accepted your (b) suggestion as my secondly.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+=========================
+
+As a simple starting point this first case study deals with the mathematical representation of signals. It is taken
+from a notebook used during a lecture. A step function
```
You mention in many places that the notebooks are taken from lectures. It's not needed here.

mwickert replied:
Deleted.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+Convolution Integral and LTI Systems
+====================================
+
+A fundamental signals and systems result states that the signal output from a *linear* and *time invariant* (LTI)
```
This sentence is awkward. The subject is plural, but the verb is singular.

mwickert replied:
I think I fixed this problem. One input signal, passing through one system to produce one output.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+.. math::
+   :label: conv_int
+
+   y(t) = \int_{-\infty}^\infty x(\lambda)h(t-\lambda)\, d\lambda \overset{\text{also}}{=}
```
Question here: What is the `also` for?

mwickert replied:
The also is to show that there are two equivalent forms of the convolution integral. I only use the second form so better for this audience just show one form.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+Students frequently have problems setting up and evaluating the convolution integral. The waveforms of interest are
+typically piecewise continuous, so the integral must be evaluated over one or more contiguous intervals. Consider the
+case of :math:`x(t) = u(t) - u(t-T)`, where :math:`u(t)` is the unit step function, and :math:`h(t) = a e^{-at}u(t)`,
+where :math:`a > 0`. To be effective in solving this problem I have students start with a sketch of the
```
"To be effective..." is awkward: you are doing the solving.

mwickert replied:
Fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+case of :math:`x(t) = u(t) - u(t-T)`, where :math:`u(t)` is the unit step function, and :math:`h(t) = a e^{-at}u(t)`,
+where :math:`a > 0`. To be effective in solving this problem I have students start with a sketch of the
+integrand :math:`h(\lambda)x(t-\lambda)`, as shown in Fig. :ref:`fig12`.
+From there you can discover the support intervals or *cases* for evaluating the integral.
```
This is another switch in narrative style. You switch from 1st person to 3rd person to 2nd person. Given the nature of the paper, it may be reasonable to switch between 1st person and 3rd person.

mwickert replied:
Reworked somewhat.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+   Solving the convolution integral in the notebook :label:`fig10`.
+
+To bring closure to the tedious analytical solution development, I encourage students to check their work. The function
```
The tedium that you are easing seems to be in checking the work, not in the solution development. From the tone of the entire paper, solution development (and exploration) is the fun part.

mwickert replied:
Did some rework.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+wired or wireless, and the signal carrying the bits may be at *baseband*, as in say Ethernet, or *bandpass* on a *carrier
+frequency*, as in WiFi. To error protect bits sent over the channel *forward error correction* (FEC) coding, such as
+*convolutional codes*, can be employed. Encoding is applied before the source bits are modulated onto the carrier to form
+the transmitted signal. With a rate 1/2 convolutional code each source bit is encoded into two channel bits. Here the specific case is taken from
```
Consider spelling out as "one-half" as you do elsewhere. If you mean it as a fraction, consider using the `\fraction` directive.

mwickert replied:
In the coding literature a literal fraction is the norm, so no change. I did change the earlier use of one-half to 1/2 to be consistent. I think this is because the code rate can be of the form p/q. The added to the developed modules section talks about rate 3/4.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+   Construction of a ``fec_conv`` object and the corresponding trellis structure for the
+   transmission of one code symbol. :label:`fig2`
+
+A *Viterbi decoder* (a form of dynamic programming) recovers the source
```
You site *Viterbi* earlier, without explaining it first. Please move this to the first place you cite it.

mwickert replied:
This has been moved up into the section that describes the code module fec_conv.py. This seems to fit better and then allows the case study to be more streamlined as most of the background is established earlier. Hopefully you agree.

dalippa replied:
I definitely agree. Both sections improved with that change.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+In the real-time digital signal processing (DSP) course C-code is written for an embedded processor. In this case the processor
+is an ARM Cortex-M4. The objective of this case study is to implement an equal-ripple *finite impulse response* (FIR)
+lowpass filter of prescribed amplitude response specifications. The filter is also LTI. Python (`scipy.signal`) is used
```
I think you need double-backticks here. `scipy.signal` shows up in italics in the PDF.

mwickert replied:
Fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+response.
+The measured frequency response is imported into the notebook using ``loadtxt()``. Fig. :ref:`fig6` compares the
+theoretical frequency response, including quantization errors, with the measured response.
+The results compare favorably. Compare theory with experiment is something students are frequently asked to do in lab
```
"Compare theory with..." is awkward and not a complete sentence.

mwickert replied:
Fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+theoretical frequency response, including quantization errors, with the measured response.
+The results compare favorably. Compare theory with experiment is something students are frequently asked to do in lab
+courses. The fact that the stopband response is not quite equal-ripple is due to coefficient
+quantization. This is easy to show right in the notebook by overlaying the frequency response using the original
```
The flow of ideas in this paragraph is very broken up. You transition from discussing the notebook, with an aside back to the classroom, and back to the notebook. Consider rewording or using a footnote to improve readability.

mwickert replied:
I changed the wording a bit.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+the system zeros and roots of :math:`D(z)` are the system poles. Students are taught that a *pole-zero*
+plot gives much insight into the frequency response of a system, in particular a filter. The module ``ssd.py`` provides
+the function ``ssd.zplane(b,a)`` where ``b`` contains the coefficients of :math:`N(z)` and ``a`` contains the
+coefficients of :math:`D(z)` (here ``a = [1]``). The even symmetry condition constrains the system zeros to lie at conjugate
```
Consider instead of "(here...", "`; in this case, a = [1]`

mwickert replied:
I accepted your suggestion.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+.. math::
+
+   \mathbf{w} = \mathbf{xy}+\mathbf{z}
```
Your display of the variables seems inconsistent. You use bold for `w` in some places but an ordinary `w` for the piecewise function. Is this intentional?

mwickert replied:
Texts mostly use caps for random variables, but one of my favorites uses boldface. This what I have done. I added some explananation on the notation.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+   One function of three random variables simulation problem. :label:`fig8`
+
+Setting up the integrations are tedious and students are timid about pushing forward with the calculus. To build
```
Instead of `integrations` do you mean `integrals`?

mwickert replied:
Good catch, fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+
+Communications and signal processing, as a discipline that sits inside electrical computer engineering, is built on
+a strong mathematical modeling foundation. Teachers and their students and researchers and their collaborators
+need tools to conquer today's
```
This sentence is awkward.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+is not only an excellent problem solving tool, but a
+means for students of all levels to explore and gain understanding of difficult engineering concepts.
+
+Today's engineering students need to be exposed to the Python/IPython and the notebook interface.
```
Why do they need to be exposed to IPython?

mwickert replied:
I have largely rewritten the conclusion and hopefully have addressed this.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,560 @@
+Today's engineering students need to be exposed to the Python/IPython and the notebook interface.
+Open source community driven alternatives should not be overlooked. A lot of credit is due to the enthusiasm of
+the scientific Python developer
+community. I am excited about the future of the IPython Notebook in the coming academic year.
```
I think the conclusion needs some more punch to it and revisit the key ideas of the paper. From my perspective, the key ideas are:
* The value of interactivity with the data: not just visual, but auditory. I don't believe the other tools you mentioned have this capability.
* Accessibility of the software: anyone can readily obtain it.
* Sharing (libraries and notebooks) to foster improved communication: between students and faculty members; between researchers, engineers, and collaborators.
* Python/IPython is a viable alternative to other tools in signal processing, but the industry has not embraced it yet

Octave *is* a community-driven alternative to MATLAB as you had cited, but it doesn't have the same enthusiasm as the Python developer community.

mwickert replied:
Bullet 1: With Mathematica you can use Manipulate to create very nice animations and play sounds. You can play sounds and write GUIs quite easily in MATLAB. There is a price tag for this.

Your last three bullets are true as I see it. I also agree with your Octave comment. Octave's primary appeal is syntax compatibility with MATLAB. I used it in consulting work prior to switching to Python. If a client want MATLAB syntax I have to goes this route or convince them why they should use Python. The Python sell is now becoming easier.

Some of your comments are now in the conclusion.

mwickert replied:
I have done another push and now with a longer conclusion section the paper jumped over 8 pages. I have tentatively commented out the Pulse Train Power Spectral Density case study. I am OK with out out it. I could also do some triming to make it fit. I will wait for editor feedback.

I also reworked three of the screenshot graphics by increasing the font size of the the overlay notes I added to the original screen shot. This should help a bit with the figure readability issue mentioned below. I am not having any readability issues myself as I generally don't try to read a PDF document at 100% unless I have a large screen and high resolution. On a mac retina display maybe.

dalippa commented:
@mwickert Re-review is complete.

A few overall comments:
* This draft is substantially better. The ideas are flowing, with the added footnotes, removed jargon, and explained acronyms. It feels much more accessible to someone with limited DSP experience, and encourages curiosity among those without it. Also, you continue to relate back to the idea that tinkering and interacting are critical parts of teaching, learning, research, and industry R&D processes. Figures seem to be much more in alignment with your citation of them.
* You are still switching tenses and narrative style (1st person, 2nd person, 3rd person) throughout the paper. Narrative style should be consistent throughout the paper; tenses should be consistent within each section, and preferably between sections.
* There still many awkward sentences. I've marked them on the actual lines to make it easier for you to fix. There are also quite a few places with grammatical errors and many sentences that might benefit from the use of a comma to improve sentence flow.
* You have a lot of text contained *inside* your figures. When viewed at 100%, they are practically unreadable. If the proceedings are printed and just digital, this might cause some readability concerns.

dalippa commented:
@mwickert I understand that at some point it must be "done", but in light of the reviewing guidelines, the paper still is quite rough - particularly in the area of grammar and clarity (overall), as well as the conclusion.

katyhuff commented:
@dalippa and @mwickert. Because there seems to be a slight tension here, I'd like to just step in briefly and first thank you both for the effort you've both put in to this process. The reworking of this document has clearly improved it and I'm very pleased to see that the remaining suggestions are nearly all very minor and grammatical in nature. (With the exception of the figure size issue - perhaps explanatory text in the captions will improve this problem). Excellent work, both.

Since genuinely *new* reviewer comments at this stage of the paper review process are indeed beyond the review deadline, I hope that the comments left here represent the full extent of what is requested and that no new suggestions will appear beyond these.

However, since narrative clarity falls within the [guidelines](https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md) we have provided to guide both authors and reviewers, it is in the best interest of the paper for suggestions to be heeded if and when they affect narrative clarity.

For all involved, I hope @mwickert will have a chance to improve clarity based on relevant suggestions given today by @dalippa as much as possible, given the time constraint. But, I also expect that @dalippa does not intend to add any geniunely new review comments, out of respect for the finite nature of time.

mwickert commented:
All,

Latest round of corrections applied. The notes are on the github page for the paper.

Mark Wickert

From: Katy Huff [mailto:notifications@github.com]
Sent: Friday, July 3, 2015 12:25 PM
To: scipy-conference/scipy_proceedings
Cc: Mark Wickert
Subject: Re: [scipy_proceedings] Paper: Signal Processing and Communications: Teaching and Research Using IPython Notebook (Wickert) (#157)


@dalippa<https://github.com/dalippa> and @mwickert<https://github.com/mwickert>. Because there seems to be a slight tension here, I'd like to just step in briefly and first thank you both for the effort you've both put in to this process. The reworking of this document has clearly improved it and I'm very pleased to see that the remaining suggestions are nearly all very minor and grammatical in nature. (With the exception of the figure size issue - perhaps explanatory text in the captions will improve this problem). Excellent work, both.

Since genuinely new reviewer comments at this stage of the paper review process are indeed beyond the review deadline, I hope that the comments left here represent the full extent of what is requested and that no new suggestions will appear beyond these.

However, since narrative clarity falls within the guidelines<https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md> we have provided to guide both authors and reviewers, it is in the best interest of the paper for suggestions to be heeded if and when they affect narrative clarity.

For all involved, I hope @mwickert<https://github.com/mwickert> will have a chance to improve clarity based on relevant suggestions given today by @dalippa<https://github.com/dalippa> as much as possible, given the time constraint. But, I also expect that @dalippa<https://github.com/dalippa> does not intend to add any geniunely new review comments, out of respect for the finite nature of time.

—
Reply to this email directly or view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/157#issuecomment-118405183>.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,580 @@
+between the lecture notes mathematics and the notebook demos. Specifically,
+text and mathematics are placed in *markdown cells*. The notebook theory is however very brief compared to that of the
+course lecture notes. Since the lecture notes are written in LaTeX, I drop the selected equations right
+in mark down cells will minimal rework. Sample calculations and simulations, with corresponding plots, are often generated
```
"Since the lecture notes are written in LaTeX..." makes no sense. Still very unclear.

mwickert replied:
Fixed: Preparing this content is easy, since the lecture notes are written in LaTeX I drop the 
selected equations right into mark down cells will minimal rework.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,580 @@
+frequency*, as in WiFi. To error protect bits sent over the channel *forward error correction* (FEC) coding, such as
+*convolutional codes*, may be employed. Encoding is applied before the source bits are modulated onto the carrier to form
+the transmitted signal. With a rate 1/2 convolutional code each source bit is encoded into two channel bits using a
+*shift register* of length :math:`K` (termed *constaint length*) with *excusive or* logic gate connections.
```
constaint? spelling error?

mwickert replied:
Yes, fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,580 @@
+use cases described in the previous sections of this paper. Case studies from industry R&D are not included here due
+to the propriety nature of the work.
+
+In all of the case studies you see that graphical results are produced using the ``pylab`` interface to
```
Still switching narrative style to 2nd person (rewrite as 1st or 3rd person)

mwickert replied:
Here is the change: In this section I present case studies that present the details on one or more of the IPython notebook
use cases described in the previous section of this paper.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,580 @@
+to learn. The waveforms of interest are
+typically piecewise continuous, so the integral must be evaluated over one or more contiguous intervals. Consider the
+case of :math:`x(t) = u(t) - u(t-T)`, where :math:`u(t)` is the unit step function, and :math:`h(t) = a e^{-at}u(t)`,
+where :math:`a > 0`. To avoid careless errors you start with a sketch of the
```
Still switching narrative style to 2nd person (rewrite as 1st or 3rd person).

mwickert replied:
Made some changes.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,580 @@
+limits. You simply need to provide an array of signal/impulse response sample values over the complete
+support interval. The screen capture of Fig. :ref:`fig11` shows how this is done in a notebook. Parameter variation is also
+explored. Seeing the two approaches agree is rewarding and a powerful testimony to how the IPython notebook improves
+learning and understanding.
```
This reworked sentence is grammatically incorrect. Is the word `agree` supposed to have been removed?

mwickert replied:
Changes to: Seeing the two approaches provide the same numerical values is rewarding and a powerful testimony to how the IPython notebook improves
learning and understanding.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,580 @@
+a strong mathematical modeling foundation. Undergraduate engineering students, despite having taken many mathematics
+courses, are often intimidated by the math they find in communications and signals processing course work.
+I cannot make the math go away, but good maodeling tools make learning and problem solving fun and exciting.
+I have found, and hopefully this paper shows you that Python/IPython, and the notebook are examples of good
```
This reworked sentence is still a bit awkward. Was the intent: `I have found, and hopefully this paper shows, that IPython notebooks are valuable mathematical modeling tools.`?

mwickert replied:
A accepted your change.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,580 @@
+The use of open-source software is increasing and cannot be overlooked in higher education. Python is readily
+accessible by anyone. It is easy to share libraries and notebooks to foster improved communication between students
+and faculty members; between researchers, engineers, and collaborators.
+IPython and the IPython notebook standout in large part due to the enthusiasm of the scientific Python developer
```
standout -> stand out

mwickert replied:
Fixed.

dalippa commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,580 @@
+The channel *signal-to-noise ratio* (SNR), also denoted :math:`E_b/N_0`) is 7 dB. Note SNR is the ratio of received
+signal power to background noise power. At a  *decision depth* of 25
+code symbols, all 16 paths merge to a common path, making it very likely that the probability of a bit error, is very
+very small. At lower a SNR it takes longer to see a traceback merge and errors bit errors are more likely.
```
"At lower a SNR it takes longer..." is awkward.

mwickert replied:
Reworked: At lower a SNR, not shown here, the increased noise level makes it take longer to see a traceback merge 
and this is indicative of an increase in the probability of making a bit error.

dalippa commented:
@mwickert The paper is substantially improved. I noted a few errors in some of the reworked sections where clarity/readability was an issue.

mwickert commented:
@dalippa , @katyhuff , completed push in response to round four comments from dalippa. Waiting for next steps.

dalippa commented:
@mwickert Therr are a few other places where you are still using 2nd person narrative, which is inconsistent and confusing.  I've mentioned this a few times now. Once those are resolved, I recommend @katyhuff and @jaberg accept the paper into the proceedings.

mwickert commented:
@dallipa and @katyhuff. I don't see the confusion you see, so I don't know how fix what appears OK to me.

dalippa commented:
I found the use of "you" as the subject in the case studies confusing;
switching between 1st (I/we), 2nd (you), and 3rd person is awkward to me.
If the proceedings chairs are ok without this change I will acquiesce.

On Saturday, July 4, 2015, mwickert <notifications@github.com> wrote:

> @dallipa and @katyhuff <https://github.com/katyhuff>. I don't see the
> confusion you see, so I don't know how fix what appears OK to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/157#issuecomment-118515996>
> .
>


-- 
David Lippa
Principal Software Engineer
david.lippa@realmassive.com
RealMassive <http://www.realmassive.com/>
t: @realmassive


* Commercial Real Estate in Real-Time*

*-----*
This message contains confidential information and is intended only for the
individual named. If you are not the named addressee you should not
disseminate, distribute or copy this e-mail. Please notify the sender
immediately by e-mail if you have received this e-mail by mistake and
delete this e-mail from your system. E-mail transmission cannot be
guaranteed to be secure or error-free as information could be intercepted,
corrupted, lost, destroyed, arrive late or incomplete, or contain viruses.
The sender therefore does not accept liability for any errors or omissions
in the contents of this message, which arise as a result of e-mail
transmission. If verification is required please request a hard-copy
version.

mwickert commented:
@dalippa and @katyhuff let me just say that when I wrote the Dummies book the editors insisted and drilled into me that I use 'I' and 'you'. This has for better or worse stuck with me over the past 3 years.

dalippa commented:
Ok. @katyhuff and @jaberg please mark as ready.

mwickert commented:
@katyhuff and @jaberg I am ready for the final step.

mwickert commented:
@katyhuff and @jaberg I was working on my presentation and noticed a spelling typo in the conclusions.

katyhuff commented:
@mwickert. Glad you fixed it. We'll probably be merging these pull requests very soon, after some final review, but until this PR is closed, you're more than welcome to make these minor changes. Thanks!
